### PR TITLE
network-generator: fixlets for vlan interfaces

### LINKED ISSUE #33954: systemd-network-generator not generating correct VLAN netdev files from command line
### systemd version the issue has been seen with

256.4-2

### Used distribution

Debian Testing

### Linux kernel version used

6.9.12-amd64

### CPU architectures issue was seen on

x86_64

### Component

systemd-network-generator

### Expected behaviour you didn't see

When using Dracut in combination with systemd-networkd module, I can set up a VLAN for ISCSI root via `vlan=enp4s0f0np0.92:enp4s0f0np0` kernel command line parameter.

### Unexpected behaviour you saw

The VLAN is not brought up because the .netdev file generated from the `vlan=` parameter lacks the VLAN ID field in the configuration, so the VLAN interface doesn't get created.

### Steps to reproduce the problem

See output below for repro with systemd-network-generator

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
> /usr/lib/systemd/systemd-network-generator --root=/tmp/t vlan=enp4s0f0np0.92:enp4s0f0np0 ip=192.168.92.11:::255.255.255.0::enp4s0f0np0.92:off:9000

> cat /tmp/t/run/systemd/network/70-enp4s0f0np0.92.netdev 
# Automatically generated by systemd-network-generator

[NetDev]
Kind=vlan
Name=enp4s0f0np0.92

> cat /tmp/t/run/systemd/network/70-enp4s0f0np0.92.network 
# Automatically generated by systemd-network-generator

[Match]
Name=enp4s0f0np0.92

[Link]
MTUBytes=9000

[Network]
DHCP=no
LinkLocalAddressing=no
IPv6AcceptRA=no

[DHCP]

[Address]
Address=192.168.92.11/24
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #33954.

### CODE REVIEW

**Review by DaanDeMeyer — APPROVED** [review]

File: src/network/generator/network-generator.c [hunk 1]
```diff
@@ -232,9 +232,24 @@ static int network_new(Context *context, const char *name, Network **ret) {
...
 Network *network_get(Context *context, const char *ifname) {
+        assert(context);
         return hashmap_get(context->networks_by_name, ifname);
 }
 
+static int network_acquire(Context *context, const char *ifname, Network **ret) {
+        Network *network;
+
+        assert(context);
+        assert(ifname);
```
Comment (bluca) [inline src/network/generator/network-generator.c:258]: assert(ret)

File: src/network/generator/network-generator.c [hunk 2]
```diff
@@ -286,9 +301,28 @@ static int netdev_new(Context *context, const char *_kind, const char *_ifname,
...
+        assert(context);
         return hashmap_get(context->netdevs_by_name, ifname);
 }
 
+static int netdev_acquire(Context *context, const char *kind, const char *name, NetDev **ret) {
+        NetDev *netdev;
+
+        assert(context);
+        assert(kind);
+        assert(name);
```
Comment (bluca) [inline src/network/generator/network-generator.c:331]: assert(ret)

File: src/network/generator/network-generator.c [hunk 3]
```diff
@@ -232,9 +233,24 @@ static int network_new(Context *context, const char *name, Network **ret) {
 }
 
 Network *network_get(Context *context, const char *ifname) {
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c]: Please update the style of return type if you touch this.

File: src/network/generator/network-generator.c [hunk 4]
```diff
@@ -974,14 +1008,25 @@ static int parse_cmdline_vlan(Context *context, const char *key, const char *val
         if (!p)
                 return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid VLAN value '%s'", value);
 
-        name = strndupa_safe(value, p - value);
+        if (!ifname_valid(p + 1))
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid VLAN physical device name '%s.'", p + 1);
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c]:
```suggestion
                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid VLAN physical device name '%s'.", p + 1);
```

File: src/network/generator/network-generator.c [hunk 5]
```diff
@@ -974,14 +1008,25 @@ static int parse_cmdline_vlan(Context *context, const char *key, const char *val
...
-                r = netdev_new(context, "vlan", name, &netdev);
-                if (r < 0)
-                        return log_debug_errno(r, "Failed to create VLAN device for '%s': %m", name);
-        }
+        name = strndup(value, p - value);
+        if (!name)
+                return log_oom_debug();
+
+        if (!ifname_valid(name))
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid VLAN device name '%s.'", name);
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c]: Ditto

File: src/network/generator/network-generator.c [hunk 6]
```diff
@@ -974,14 +1008,25 @@ static int parse_cmdline_vlan(Context *context, const char *key, const char *val
...
+        if (!ifname_valid(p + 1))
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid VLAN physical device name '%s.'", p + 1);
 
-        netdev = netdev_get(context, name);
-        if (!netdev) {
-                r = netdev_new(context, "vlan", name, &netdev);
-                if (r < 0)
-                        return log_debug_errno(r, "Failed to create VLAN device for '%s': %m", name);
-        }
+        name = strndup(value, p - value);
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c]: Hmm, why not use extract_first_word then?

**Comment by yuwata** [comment]
The last commit is slightly big. Please re-review.

File: src/network/generator/network-generator.c [hunk 7]
```diff
@@ -285,11 +317,32 @@ static int netdev_new(Context *context, const char *_kind, const char *_ifname,
...
+        assert(context);
+        assert(kind);
+        assert(name);
+
+        netdev = netdev_get(context, name);
+        if (!netdev)
+                return netdev_new(context, kind, name, ret);
+
+        if (streq_ptr(netdev->kind, kind))
+                return -EEXIST; /* conflicting netdev already exists. */
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c:338]: Hmm, the condition seems inverted?

Comment (yuwata) [reply src/network/generator/network-generator.c:338]: Ouch.

File: src/network/generator/network-generator.c [hunk 8]
```diff
@@ -566,296 +624,268 @@ static int network_set_bond(Context *context, const char *ifname, const char *va
...
 
-        p = strchr(value, ':');
-        if (!p)
-                mtu = value;
-        else
-                mtu = strndupa_safe(value, p - value);
-
-        r = network_set_mtu(context, ifname, mtu);
-        if (r < 0)
+        p = value;
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c:648]: Just initialize `p` right away above?

Comment (YHNdnzj) [inline src/network/generator/network-generator.c:648]: Just initialize `p` right away above?

Comment (YHNdnzj) [inline src/network/generator/network-generator.c]: Just initialize `p` right away above?

Comment (YHNdnzj) [inline src/network/generator/network-generator.c:648]: Just initialize `p` right away above?

Comment (YHNdnzj) [inline src/network/generator/network-generator.c]: It's fine to use `value` directly?

File: src/network/generator/network-generator.c [hunk 9]
```diff
@@ -566,296 +624,268 @@ static int network_set_bond(Context *context, const char *ifname, const char *va
...
+        r = extract_ip_address_str(family, &p, &buf);
+        if (r < 0)
+                return r;
 
-                *value = p + 1;
+        if (isempty(buf)) {
+                *ret = IN_ADDR_NULL;
+                r = 0;
+        } else {
+                k = in_addr_from_string(family, buf, ret);
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c:717]: Maybe `assert(r > 0)` here?

File: src/network/generator/network-generator.c [hunk 10]
```diff
@@ -884,10 +914,11 @@ static int parse_cmdline_ip(Context *context, const char *key, const char *value
 }
 
 static int parse_cmdline_rd_route(Context *context, const char *key, const char *value) {
-        union in_addr_union addr = {}, gateway = {};
+        _cleanup_free_ char *buf = NULL;
+        union in_addr_union dest, gateway;
         unsigned char prefixlen;
-        const char *buf, *p;
         int family, r;
+        const char *p;
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c]: I would like to place `r` at last, but doesn't matter.

File: src/network/generator/network-generator.c [hunk 11]
```diff
@@ -959,60 +971,76 @@ static int parse_cmdline_rd_peerdns(Context *context, const char *key, const cha
...
         NetDev *netdev;
         int r;
 
         assert(context);
         assert(key);
 
+        /* vlan=<vlanname>:<phydevice> */
+
         if (proc_cmdline_value_missing(key, value))
                 return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Missing value for '%s'", key);
```
Comment (YHNdnzj) [inline src/network/generator/network-generator.c]: Not your fault. But `proc_cmdline_value_missing` logs at warning level on its own. All uses in this file seems inappropriate?

Comment (YHNdnzj) [reply src/network/generator/network-generator.c]: I.e. either remove the duplicated debug log or step away from `proc_cmdline_value_missing`?

**Review by YHNdnzj — APPROVED** [review]
LGTM, just some minor stuff

**Comment by yuwata** [comment]
@YHNdnzj Thank you for the review. Addressed comments, and upgraded the green label.

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
I backported the following:
* 202651f965 network-generator: drop wrong warning for rd.peerdns without value
* a9a330cb84 network-generator: ignore kernel command line without value
* 13c51806ba network-generator: parse vlan ID from vlan interface name
* ad0d482c11 network-generator: allow to create bridge or friends without slave interfaces
* e7ef37336a network-generator: vlan= can be specified multiple times

The rest is bigger and seems risky for backports.

**Pull request merged**